haystack/tools/component_tool.py

```diff
@@ -0,0 +1,330 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import fields, is_dataclass
+from inspect import getdoc
+from typing import Any, Callable, Dict, Optional, Union, get_args, get_origin
+
+from pydantic import TypeAdapter
+
+from haystack import logging
+from haystack.core.component import Component
+from haystack.core.serialization import (
+    component_from_dict,
+    component_to_dict,
+    generate_qualified_class_name,
+    import_class_by_name,
+)
+from haystack.lazy_imports import LazyImport
+from haystack.tools import Tool
+from haystack.tools.errors import SchemaGenerationError
+
+with LazyImport(message="Run 'pip install docstring-parser'") as docstring_parser_import:
+    from docstring_parser import parse
+
+
+logger = logging.getLogger(__name__)
+
+
+class ComponentTool(Tool):
+    """
+    A Tool that wraps Haystack components, allowing them to be used as tools by LLMs.
+
+    ComponentTool automatically generates LLM-compatible tool schemas from component input sockets,
+    which are derived from the component's `run` method signature and type hints.
+
+
+    Key features:
+    - Automatic LLM tool calling schema generation from component input sockets
+    - Type conversion and validation for component inputs
+    - Support for types:
+        - Dataclasses
+        - Lists of dataclasses
+        - Basic types (str, int, float, bool, dict)
+        - Lists of basic types
+    - Automatic name generation from component class name
+    - Description extraction from component docstrings
+
+    To use ComponentTool, you first need a Haystack component - either an existing one or a new one you create.
+    We can create a ComponentTool from the component by passing the component to the ComponentTool constructor.
```

Comment:
Found one more improvement

```suggestion
    You can create a ComponentTool from the component by passing the component to the ComponentTool constructor.
```